pkg/cmd/podidentity/detect.go

```diff
@@ -0,0 +1,410 @@
+			return detectCmd.prerun()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return detectCmd.run()
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVar(&detectCmd.namespace, "namespace", "default", "Namespace to detect the configuration")
+	f.StringVarP(&detectCmd.outputDir, "output-dir", "o", "", "Output directory to write the configuration files")
+	f.IntVarP(&detectCmd.proxyPort, "proxy-port", "p", 8000, "Proxy port to use for the proxy container")
+
+	_ = cmd.MarkFlagRequired("output-dir")
+
+	return cmd
+}
+
+func (dc *detectCmd) prerun() error {
+	var err error
+	dc.serializer = json.NewSerializerWithOptions(
+		json.DefaultMetaFactory, scheme, scheme,
+		json.SerializerOptions{
+			Yaml:   true,
+			Pretty: true,
+			Strict: true,
+		},
+	)
+	dc.kubeClient, err = kuberneteshelper.GetKubeClient()
+	if err != nil {
+		return err
+	}
+	// create output directory if it doesn't exist
+	if _, err := os.Stat(dc.outputDir); os.IsNotExist(err) {
+		return os.MkdirAll(dc.outputDir, 0755)
+	}
+
+	return nil
+}
+
+func (dc *detectCmd) run() error {
+	log.Debugf("detecting aad-pod-identity configuration in namespace: %s", dc.namespace)
+
+	// Implementing force namespaced mode
+	// 1. Get AzureIdentityBinding in the namespace
+	// 2. Get AzureIdentity referenced by AzureIdentityBinding and store in map with aadpodidbinding label value as key and AzureIdentity as value
+	// 3. Get all pods in the namespace that have aadpodidbinding label
+	// 4. For each pod, check if there is an owner reference (deployment, statefulset, cronjob, job, daemonset, replicaset, replicationcontroller)
+	// 5. If there is an owner reference, get the owner reference object and add to map with aadpodidbinding label value as key and owner reference as value
+	// 6. If no owner reference, then assume it's a static pod and add to map with aadpodidbinding label value as key and pod as value
+	// 7. Loop through the first map and generate new config file for each owner reference and service account
+	//    1. If owner is using a service account, get the service account and generate a config file with it
+	//    2. If owner doesn't use service account, generate a new service account yaml file with owner name as service account name
+
+	azureIdentityBindings, err := kuberneteshelper.ListAzureIdentityBinding(context.TODO(), dc.kubeClient, dc.namespace)
+	if err != nil {
+		return err
+	}
+	azureIdentities, err := kuberneteshelper.ListAzureIdentity(context.TODO(), dc.kubeClient, dc.namespace)
+	if err != nil {
+		return err
+	}
+
+	labelsToAzureIdentityMap := make(map[string]aadpodv1.AzureIdentity)
+	for _, azureIdentityBinding := range azureIdentityBindings {
+		if azureIdentityBinding.Spec.Selector == "" || azureIdentityBinding.Spec.AzureIdentity == "" {
+			continue
+		}
+		if azureIdentity, ok := azureIdentities[azureIdentityBinding.Spec.AzureIdentity]; ok {
+			labelsToAzureIdentityMap[azureIdentityBinding.Spec.Selector] = azureIdentity
```

Comment:
How are we handling two different AzureIdentityBindings with the same selector?

Reply:
Good question! I've updated `ListAzureIdentityBinding` to first sort the list of `AzureIdentityBinding` and then add to map if not exists. We'll annotate the service account with the clientID from the first `AzureIdentityBinding` -> `AzureIdentity` -> clientID (this behavior is deterministic with the sort) as we can only annotate with 1 clientID.

If the workload is using multiple pod identities with AAD Pod Identity, they will follow the similar model with workload identity. The `AZURE_CLIENT_ID` env var will be just one of the client ids. The workload will most likely not use that env var and instead rely on runtime variables they have configured for which managed identity to use to access which azure resource.

Hope this clarifies your question!

Reply:
Added a comment with the same explanation!

Reply:
`azureIdentityBindings` is a map so iterating it is still indeterministic right? We need to store `azureIdentityBindings` in a slice to preserve the ordering.

Reply:
good catch! updated it in https://github.com/Azure/azure-workload-identity/pull/432/commits/5f129a09753fd17f66eb7b5e771c86d806ffd77d and added some unit tests